### DISCUSSION TITLE
Similar tool for object storage (AWS S3, GCS, Azure Storage)?

### DISCUSSION THREAD

**Discussion by p-hebert** [theory/sqitch#647] [Q&A]
Hello!

A bit off topic for this discussion section, but I'd like to know if there is a change management system like sqitch for storage objects that allows to declaratively write migrations that can be tracked, reverted and verified.

The goal is to make our change in bucket directory structure predictable, documented and revert-able (where possible).

Anyone knows a technology that does this specifically?

Thanks!

**Comment by theory** [comment]
[Terraform](https://www.terraform.io)?

**Reply by p-hebert** [reply]
Hey David!

I'm not sure how I'd go around doing that. Can I use Terraform to move an existing object to a different destination? What about applying a script that converts the file type (e.g. from csv to parquet) across all objects?
My understanding of Terraform is that it's more for ensuring state than for transforming state.

The model I'm trying to apply is very similar to that of sqitch. Given the assumption that _a table_ in SQL land == _a set of object matching a path pattern in at least one storage bucket_ in storage land, these are use-cases that translate to each other:

- Adding a column to a table == Adding metadata key to all objects matching a path pattern
- Updating data in a column of a table == Updating metadata key to all objects matching a path pattern
- Reindexing an PK column == Renaming object (or moving objects) matching a path pattern
- Updating a binary column of a table == Applying a transformation on the content of an object for all objects matching a path pattern

Do you still think that Terraform applies to this use-case?

**Reply by theory** [reply]
I honestly don't know, sorry, I'm a Terraform newb myself.

**Reply by p-hebert** [reply]
Thanks for attempting to answer anyways :)
